File: typescript/src/pagination.ts
/**
 * Cursor pagination — the same wire format the Rust and Dart cores produce.
 *
 * An offset is wrong for a feed that grows while you read it. The cursor
 * encodes the position of the last row returned, base64url of
 * `v1:<sort-key>:<id>`, so the next page continues from a fixed point.
 */

export const MAX_PAGE_LIMIT = 100;
export const DEFAULT_PAGE_LIMIT = 25;

const CURSOR_VERSION = "v1";

export class PaginationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PaginationError";
  }
}

export interface Cursor {
  readonly sortKey: string;
  readonly id: string;
}

export interface PageRequest {
  readonly limit: number;
  readonly after: Cursor | null;
}

export interface Page<T> {
  readonly items: readonly T[];
  readonly next: Cursor | null;
}

/** Build a cursor. A `:` in either field would make the encoding ambiguous. */
export function cursor(sortKey: string, id: string): Cursor {
  if (sortKey.length === 0 || id.length === 0) {
    throw new PaginationError("a cursor needs both a sort key and an id");
  }
  if (sortKey.includes(":") || id.includes(":")) {
    throw new PaginationError("a cursor field may not contain ':'");
  }
  return { sortKey, id };
}

export function encodeCursor(value: Cursor): string {
  return base64UrlEncode(
    new TextEncoder().encode(`${CURSOR_VERSION}:${value.sortKey}:${value.id}`),
  );
}

/** Parse a cursor this API issued. Anything else throws rather than guessing. */
export function decodeCursor(encoded: string): Cursor {
  const text = new TextDecoder("utf-8", { fatal: true }).decode(
    base64UrlDecode(encoded),
  );
  // `split` with a limit drops the tail; slice the remainder back on instead so
  // a legitimate id can never be silently truncated.
  const parts = text.split(":");
  if (parts.length < 3) {
    throw new PaginationError("cursor is not a cursor this API issued");
  }
  const [version, sortKey, ...rest] = parts;
  if (version !== CURSOR_VERSION) {
    throw new PaginationError("cursor is not a cursor this API issued");
  }
  return cursor(sortKey ?? "", rest.join(":"));
}

export function pageRequest(
  limit: number = DEFAULT_PAGE_LIMIT,
  after: Cursor | null = null,
): PageRequest {
  if (!Number.isInteger(limit) || limit < 1 || limit > MAX_PAGE_LIMIT) {
    throw new PaginationError(`page limit ${limit} is outside 1..=${MAX_PAGE_LIMIT}`);
  }
  return { limit, after };
}

/** Query-string fragment, with no leading `?` so a caller can compose it. */
export function pageQuery(request: PageRequest): string {
  const parts = [`limit=${request.limit}`];
  if (request.after !== null) {
    parts.push(`after=${encodeCursor(request.after)}`);
  }
  return parts.join("&");
}

export function isLastPage<T>(page: Page<T>): boolean {
  return page.next === null;
}

const ALPHABET =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

/** Unpadded base64url (RFC 4648 §5). No dependency, works in every runtime. */
export function base64UrlEncode(bytes: Uint8Array): string {
  let out = "";
  for (let index = 0; index < bytes.length; index += 3) {
    const b0 = bytes[index] ?? 0;
    const b1 = bytes[index + 1];
    const b2 = bytes[index + 2];
    const triple = (b0 << 16) | ((b1 ?? 0) << 8) | (b2 ?? 0);
    out += ALPHABET[(triple >> 18) & 0x3f];
    out += ALPHABET[(triple >> 12) & 0x3f];
    if (b1 !== undefined) {
      out += ALPHABET[(triple >> 6) & 0x3f];
    }
    if (b2 !== undefined) {
      out += ALPHABET[triple & 0x3f];
    }
  }
  return out;
}

/** Rejects padding, whitespace and foreign characters rather than skipping them. */
export function base64UrlDecode(input: string): Uint8Array {
  if (input.length === 0) {
    throw new PaginationError("empty base64url input");
  }
  const out: number[] = [];
  let buffer = 0;
  let bits = 0;
  for (const character of input) {
    const value = ALPHABET.indexOf(character);
    if (value < 0) {
      throw new PaginationError(`invalid base64url character ${JSON.stringify(character)}`);
    }
    buffer = (buffer << 6) | value;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.push((buffer >> bits) & 0xff);
    }
  }
  if (bits >= 6 || (buffer & ((1 << bits) - 1)) !== 0) {
    throw new PaginationError("truncated base64url input");
  }
  return Uint8Array.from(out);
}

File: dart/lib/src/pagination.dart
import 'dart:convert';

import 'errors.dart';

/// Cursor pagination — the same wire format the Rust and TypeScript cores use.
const int maxPageLimit = 100;
const int defaultPageLimit = 25;

const String _cursorVersion = 'v1';
const String _alphabet =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';

/// An opaque continuation token: base64url of `v1:<sortKey>:<id>`.
final class Cursor {
  Cursor(this.sortKey, this.id) {
    if (sortKey.isEmpty || id.isEmpty) {
      throw const InvalidCursor('a cursor needs both a sort key and an id');
    }
    if (sortKey.contains(':') || id.contains(':')) {
      throw const InvalidCursor("a cursor field may not contain ':'");
    }
  }

  /// Parse a cursor this API issued. Anything else throws.
  factory Cursor.decode(String encoded) {
    final bytes = base64UrlDecodeStrict(encoded);
    final String text;
    try {
      text = const Utf8Decoder().convert(bytes);
    } on FormatException {
      throw const InvalidCursor('payload is not valid UTF-8');
    }
    final parts = text.split(':');
    if (parts.length < 3 || parts[0] != _cursorVersion) {
      throw const InvalidCursor('unrecognised payload');
    }
    return Cursor(parts[1], parts.sublist(2).join(':'));
  }

  final String sortKey;
  final String id;

  String encode() =>
      base64UrlEncodeUnpadded(utf8.encode('$_cursorVersion:$sortKey:$id'));

  @override
  bool operator ==(Object other) =>
      other is Cursor && other.sortKey == sortKey && other.id == id;

  @override
  int get hashCode => Object.hash(sortKey, id);

  @override
  String toString() => 'Cursor($sortKey, $id)';
}

/// A request for one page.
final class PageRequest {
  PageRequest({this.limit = defaultPageLimit, this.after}) {
    if (limit < 1 || limit > maxPageLimit) {
      throw PageLimitOutOfRange(limit, maxPageLimit);
    }
  }

  final int limit;
  final Cursor? after;

  /// Query-string fragment with no leading `?`, so a caller can compose it.
  String toQuery() {
    final cursor = after;
    if (cursor == null) {
      return 'limit=$limit';
    }
    return 'limit=$limit&after=${cursor.encode()}';
  }
}

/// One page of results plus the cursor that continues it.
final class Page<T> {
  const Page(this.items, this.next);

  final List<T> items;
  final Cursor? next;

  bool get isLast => next == null;
  bool get isEmpty => items.isEmpty;
  int get length => items.length;
}

/// Unpadded base64url (RFC 4648 §5).
///
/// Dart's `base64Url` codec pads; the wire format does not, and a padded cursor
/// would not match the Rust or TypeScript encoders.
String base64UrlEncodeUnpadded(List<int> input) {
  final buffer = StringBuffer();
  for (var index = 0; index < input.length; index += 3) {
    final b0 = input[index];
    final b1 = index + 1 < input.length ? input[index + 1] : null;
    final b2 = index + 2 < input.length ? input[index + 2] : null;
    final triple = (b0 << 16) | ((b1 ?? 0) << 8) | (b2 ?? 0);
    buffer.write(_alphabet[(triple >> 18) & 0x3f]);
    buffer.write(_alphabet[(triple >> 12) & 0x3f]);
    if (b1 != null) {
      buffer.write(_alphabet[(triple >> 6) & 0x3f]);
    }
    if (b2 != null) {
      buffer.write(_alphabet[triple & 0x3f]);
    }
  }
  return buffer.toString();
}

/// Rejects padding, whitespace and foreign characters rather than skipping them.
List<int> base64UrlDecodeStrict(String input) {
  if (input.isEmpty) {
    throw const InvalidCursor('empty base64url input');
  }
  final out = <int>[];
  var buffer = 0;
  var bits = 0;
  for (final unit in input.codeUnits) {
    final value = _alphabet.codeUnits.indexOf(unit);
    if (value < 0) {
      throw const InvalidCursor('invalid base64url character');
    }
    buffer = (buffer << 6) | value;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out.add((buffer >> bits) & 0xff);
    }
  }
  if (bits >= 6 || (buffer & ((1 << bits) - 1)) != 0) {
    throw const InvalidCursor('truncated base64url input');
  }
  return out;
}
